### PR TITLE
Added: Recovery counter, flow change, changeGuardianParams issue

### PR DESCRIPTION
**Pull request by filmakarov**
# Summary

## changeGuardianParams issue
Fixed issue with 'changeGuardianParams' could be added by SA to also add guardians leading to inconsistency of guardiansCount.

## Recoveries limitation
Added `recoveriesLeft` counter. Now, only user-defined # of recoveries can happen. After this, no userOps can be validated, so the user may reset the module and set new guardians.
To achieve this, it is required to change the flow of how recovery request is executed:
- Was: EP => SA.execute => recoverer.RecoveryMethod
- Now: EP => SA.execute => AccRecoveryModule.executeRecovery => SA.execTransactionFromModuleReturnData => recoverer.RecoveryMethod
Due to introducing `AccRecoveryModule.executeRecovery` method, we have a single point of handling the execution of the recovery requests.
Here the module:
- Checks if the recovery has been properly executed
- Emits event
- Clears the request
- Decrements `recoveriesLeft`

Also, the module now checks that the recovery request uses `executeRecovery`. Requests that do not use `executeRecovery` are not allowed.

## Reset Module
`resetModule` method was introduced.
It still requires to provide guardians from off-chain. Otherwise, it would require storing them in a mapping, which is ugly.
In any case, all the guardians are available through events.

## Tests
The according changes were made to the test suite.

Related Issue: #[SMA-28](https://linear.app/biconomy/issue/SMA-28/account-recovery-kawach)

## Change Type

- [x] Refactor
- [x] New Feature
- [x] Breaking Change

# Checklist

- [x] My code follows this project's style guidelines
- [x] I've reviewed my own code
- [x] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [x] My changes generate no new warnings
- [x] I've added tests that prove my fix is effective or my feature works
- [x] All unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

### CODE REVIEW

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (Aboudjem) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:201]: Fetching both `validAfter` and `validUntil` in a single call can reduce gas.
```solidity
TimeFrame memory guardianTimeFrame = _guardians[currentGuardian][smartAccount];

uint48 validAfter = guardianTimeFrame.validAfter;
uint48 validUntil = guardianTimeFrame.validUntil;
```

Please double check

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (Aboudjem) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:308]: I think it would be beneficial to split this function into 2-3 subfunctions. Having ~200 lines for a function makes it hard to read, especially with all these conditions and assembly code.
Something like:
- checkRecoveryStatus
- validateSignatures
- verifyRecoveryRequest

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (Aboudjem) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:1]: the contract lacks a check to prevent the same guardian from being added multiple times. This oversight could lead to a situation wher if a user loses access to its smart account and there are duplicate guardians recovery becomes impossible

**Review by Aboudjem — CHANGES_REQUESTED** [review]

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:87]: if (recoveryThreshold > length) it reverts => recoveryThreshold is <= length so in theory it can be 0

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:201]: >Fetching both validAfter and validUntil in a single call can reduce gas.

like this?

```
(uint48 validAfter, uint48 validUntil) = (
                _guardians[currentGuardian][smartAccount].validAfter, 
                _guardians[currentGuardian][smartAccount].validUntil
            );
```

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:1]: Can you elaborate please?

I think

```
if (_guardians[guardians[i]][msg.sender].validUntil != 0)
                revert GuardianAlreadySet(guardians[i], msg.sender);
```
this check in `initForSmartAccount` and `addGuardian` methods should do the job of protecting against it

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol]: Thank you for inspiring me to do this :) Now validateUserOp looks much prettier and easier to understand

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (Aboudjem) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:85]: my bad, I forgot the `revert`, I read it as a `require` 🫥

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (Aboudjem) [reply contracts/smart-account/modules/AccountRecoveryModule.sol]: Not sure, the goal is to access to storage once here
that's why the `TimeFrame memory guardianTimeFrame = _guardians[currentGuardian][smartAccount];`

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (Aboudjem) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:1]: No I was referring to this:

```solidity
            // gas efficient way to ensure all guardians are unique
            // requires from dapp to sort signatures before packing them into bytes
            if (currentGuardianAddress <= lastGuardianAddress)
                revert("AccRecovery: NotUnique/BadOrder");
```

**Review by Aboudjem — APPROVED** [review]
LGTM just please double check this: 

https://github.com/bcnmy/scw-contracts/pull/185/files/96b7e78bdc03c0c27d03cdca95c0368c784a7a76..93da5ada9c403a808117064ce3204b7057a87a71#r1434612266

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol]: got it. yep, it saves about 150 gas. fixed, thnx!

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:1]: This part is not in charge of adding guardians, it is to ensure all the guardians who signed the userOp are unique (and sorted). If they are not unique, it will revert with "NotUnique/BadOrder"

Comment (Aboudjem) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:1]: Yes, that's my point! my concern is about the possibility of identical guardians being added. If two identical guardians are added to the guardian list, it could potentially make the recovery impossible, and this problem will be seen only while the user will try to recover

Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:1]: I see. However, adding two identical guardians is not possible because of this
```
if (_guardians[guardian][msg.sender].validUntil != 0)
            revert GuardianAlreadySet(guardian, msg.sender);
```
So everything should be fine

File: contracts/smart-account/interfaces/IAccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/interfaces/IAccountRecoveryModule.sol:46]: can index 2 more values depending on the importance

File: contracts/smart-account/interfaces/IAccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/interfaces/IAccountRecoveryModule.sol:122]: can index both

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:134]: can rename to AccountRecovery

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:165]: can rename to something more readable like isUserOpValidForRecoveryRequestSubmission

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:309]: how would one procure guarding hashes?

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:444]: can move this effects later after "checks"

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:448]: can leave second empty if not using retData

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:564]: if we're using codes can we use them everywhere

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol:608]: why * 2 ?

File: test/bundler-integration/module/AccountRecovery.Module.specs.ts
Comment (livingrockrises) [inline test/bundler-integration/module/AccountRecovery.Module.specs.ts:199]: can you make any transaction using executeRecovery or only to one of validation modules / just ECDSA?

**Review by livingrockrises — COMMENTED** [review]
reviewed

**Comment by livingrockrises** [comment]
what is defined as recoverer.RecoveryMethod?

**Review by Aboudjem — APPROVED** [review]

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:309]: ask a guardian (person) to sign the `CONTROL_MESSAGE`, then hash the signature

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:444]: due to the re-entrancy potential, I decided to keep effect (remove request and decrease recoveriesLeft) before interaction (execute request)

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:448]: retData is used in the `revert` next

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (filmakarov) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:608]: because `moduleSignature` contains signatures over `userOpHash` and signatures over `CONTROL_MESSAGE`.
latter is used to recover guardian's address and compare it with the address that signed `userOpHash`.

File: test/bundler-integration/module/AccountRecovery.Module.specs.ts
Comment (filmakarov) [reply test/bundler-integration/module/AccountRecovery.Module.specs.ts:199]: any transaction

**Comment by filmakarov** [comment]
do not merge please, may need slight improvement before merging as per Zellic notes

**Comment by livingrockrises** [comment]
> do not merge please, may need slight improvement before merging as per Zellic notes

let us know here once all improvements are pushed.

**Comment by filmakarov** [comment]
> > do not merge please, may need slight improvement before merging as per Zellic notes
> 
> let us know here once all improvements are pushed.

all improvements are pushed, waiting for the confirmation from Zellic that all good

**Comment by filmakarov** [comment]
@livingrockrises can you review and approve please?

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [inline contracts/smart-account/modules/AccountRecoveryModule.sol]: please remove hardhat console.sol and console logs

File: contracts/smart-account/modules/AccountRecoveryModule.sol
Comment (livingrockrises) [reply contracts/smart-account/modules/AccountRecoveryModule.sol:461]: hmm and any checks should come before those both. 
we can also add reentrancy guard as base and modifier nonReentrant if it makes sense

File: test/bundler-integration/module/AccountRecovery.Module.specs.ts
Comment (livingrockrises) [reply test/bundler-integration/module/AccountRecovery.Module.specs.ts:199]: I see. this can be used to lock the smart account if we introduce such feature at implementation level. recently saw a guardian system that helps to lock the account (besides doing the recovery) [ can be helpful in the cases when your private key is stolen and not lost ]

**Pull request merged**
